firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -159,66 +163,101 @@ public static FirebaseInstallations getInstance(@NonNull FirebaseApp app) {
+  private Task<InstallationTokenResult> addGetAuthTokenListener(
+      @AuthTokenOption int authTokenOption) {
+    TaskCompletionSource<InstallationTokenResult> taskCompletionSource =
+        new TaskCompletionSource<>();
+    StateListener l = new GetAuthTokenListener(utils, taskCompletionSource);
+    synchronized (lock) {
+      if (authTokenOption == FORCE_REFRESH) {
+        shouldRefreshAuthToken = true;
+      }
+      listeners.add(l);
+    }
+    return taskCompletionSource.getTask();
   }
 
-  @NonNull
-  private static <F, T> Continuation<F, T> orElse(@NonNull Supplier<T> supplier) {
-    return t -> {
-      if (t.isSuccessful()) {
-        return (T) t.getResult();
+  private void triggerOnStateReached(PersistedFidEntry persistedFidEntry) {
+    synchronized (lock) {
+      Iterator<StateListener> it = listeners.iterator();
+      while (it.hasNext()) {
+        StateListener l = it.next();
+        boolean doneListening = l.onStateReached(persistedFidEntry);
+        if (doneListening) {
+          it.remove();
+        }
       }
-      return supplier.get();
-    };
+    }
   }
 
-  @NonNull
-  private static <F, T> Continuation<F, T> awaitFidRegistration(
-      @NonNull Supplier<T> supplier, AwaitListener listener) {
-    return t -> {
-      // Waiting for Task that registers FID on the FIS Servers
-      listener.await(AWAIT_TIMEOUT_IN_SECS, TimeUnit.SECONDS);
-      return supplier.get();
+  private final Runnable doRegistration() {
+    return () -> {
+      try {
+        PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
+
+        // New FID needs to be created
+        if (persistedFidEntry.isErrored() || persistedFidEntry.isNotGenerated()) {
+          String fid = utils.createRandomFid();
+          persistFid(fid);
+          persistedFidEntry = persistedFid.readPersistedFidEntryValue();
+        }
+
+        // Notify the listeners only after force refreshing auth token if shouldRefreshAuthToken is
+        // true
+        synchronized (lock) {
+          if (!shouldRefreshAuthToken) {
+            triggerOnStateReached(persistedFidEntry);
+          }
+        }
+
+        // FID needs to be registered
+        if (persistedFidEntry.isUnregistered()) {
+          registerAndSaveFid(persistedFidEntry);
+          persistedFidEntry = persistedFid.readPersistedFidEntryValue();
+        }
+
+        // Don't notify the listeners at this point; we might as well make ure the auth token is up
+        // to date before letting them know.
+
+        boolean needRefresh = utils.isAuthTokenExpired(persistedFidEntry);
+        if (!needRefresh) {
+          synchronized (lock) {
+            needRefresh = shouldRefreshAuthToken;
+          }
+        }
+
+        // Refresh Auth token if needed
+        if (needRefresh) {
+          synchronized (lock) {
+            shouldRefreshAuthToken = false;
+          }
+          fetchAuthTokenFromServer(persistedFidEntry);
```

Comment:
why do the fetch after setting `shouldRefreshAuthToken` to false instead of in reverse?

Reply:
Updated.